### ISSUE TITLE
Intermittent error: UNAUTHENTICATED: Request had invalid authentication credentials.

### ISSUE THREAD

**Issue by karthikv** [googleapis/nodejs-speech#243]
We've been using google cloud speech alongside these nodejs client bindings in production for many months now. Intermittently, we get the following error:

```
Error: 16 UNAUTHENTICATED: Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.

at Object.exports.createStatusError (node_modules/grpc/src/common.js line 87 col 15)
at ClientDuplexStream._emitStatusIfDone (node_modules/grpc/src/client.js line 235 col 26)
at ClientDuplexStream._receiveStatus (node_modules/grpc/src/client.js line 213 col 8)
at Object.onReceiveStatus (node_modules/grpc/src/client_interceptors.js line 1290 col 15)
at InterceptingListener._callNext (node_modules/grpc/src/client_interceptors.js line 564 col 42)
at InterceptingListener.onReceiveStatus (node_modules/grpc/src/client_interceptors.js line 614 col 8)
```

In the last 30 days, we've received ~100 of these errors from about over 40,000 streaming recognize sessions, so this only happens in about ~0.25% of cases. In all the remaining cases, the streaming recognize works as expected. Note that this problem has been going on since we started using this library, not just for the past month.

#### Environment details

  - OS: debian jessie on GCE
  - Node.js version: 10.13.0
  - npm version: 6.4.1
  - `@google-cloud/speech` version: 2.0.0

#### Steps to reproduce
There's not a clear reproduction case--this error comes about naturally after running streaming recognizes for some extended period of time. My guess here is that the access tokens are sometimes not refreshed in a timely manner, or perhaps some clock skew is causing them to be invalid. 

Here are some issues I've found in the python client libraries that resemble the problem I'm seeing, in case this is useful:

- https://github.com/googleapis/google-cloud-python/issues/5014
- https://github.com/googleapis/google-cloud-python/issues/2684
- https://github.com/googleapis/google-cloud-python/issues/3259

I'll definitely try upgrading to the latest library version as a first step.

**Comment by karthikv** [comment]
Just want to follow up here: any idea if/when this will be looked into? We're heavy users of the speech API, and we're encountering this 30+ times a day, which is a poor experience for our users. Upgrading to 2.2.0 doesn't fix this issue.

**Comment by JustinBeckwith** [comment]
Greetings!  What execution environment are you running in?  Cloud Functions?  App Engine?  Local?

**Comment by karthikv** [comment]
We're seeing this error on GCE using debian jessie. The first post has a few more details about the environment.

**Comment by JustinBeckwith** [comment]
Apologies for missing that.  Are you relying application default authentication, or are you packaging a keyfile along with your sources?  If you would be willing - I'd love to know if this reproduces with a local keyfile instead of relying on the GCE metadata server. May give us a clue to what's happening here.

**Comment by karthikv** [comment]
@JustinBeckwith I'm currently using a keyfile.

**Comment by JustinBeckwith** [comment]
Ohhhkay that's wild.  Are you using other services by chance?  I'm curious if this is something specific to the speech API, or something you're seeing elsewhere.

Adding @theacodes as she may have insights here as well.

**Comment by theacodes** [comment]
We're not seeing this in other languages, so I'm hesitant to think it's a server or frontend issue. 

I'm wondering a few things:

1. Does this library eagerly refresh credentials if `expire_time < now - clock_skew_buffer`?
2. How long are these streaming requests?
3. How long are the clients your using hanging around?
4. Is it possible that the credentials expire during the stream and the service then closes the stream with unauthenticated?

**Comment by callmehiphop** [comment]
@sduskis @ajaaym just an FYI, this sounds pretty similar to what we're seeing in PubSub.

**Comment by sduskis** [comment]
@callmehiphop, I agree.  Streaming RPCs need some sort of retry on UNAUTHORIZED, if they've been open for a while. 

https://github.com/googleapis/nodejs-pubsub/issues/318 showed similar issues, and it references this same behavior in different languages.

**Comment by JustinBeckwith** [comment]
Bringing in @jadekler for the retry grpc discussion 🙃

**Comment by jadekler** [comment]
As @theacodes alluded to in https://github.com/googleapis/nodejs-speech/issues/243#issuecomment-459553924 (question #1), I expect not to need retry on UNAUTHORIZED as long as the auth library is refreshing credentials before the deadline.

Is that not the case in python/java auth? Or, perhaps I'm missing something.

(I defer to @theacodes and @broady who know more about auth and auth refresh than I do, though)

**Comment by sduskis** [comment]
@jadekler, long running streaming APIs have interesting failure modes relating to credentials.  The Token is passed with the RPC is started, but it never changes even when the auth token expires.

Let's say the auth token is 50 minutes old, and a new streaming RPC is started.  If the RPC is open past the auth expiration, the service may send UNAUTHORIZED.

**Comment by jadekler** [comment]
That makes sense, but it doesn't seem like we need that in Go. I guess I don't really understand why. @broady do you know how this stuff works? cc @dfawley

**Comment by jadekler** [comment]
@dfawley mentions it's server-implementation specific. It's definitely plausible that a server might just check once - at the start of the RPC (the long-lived stream). I naively expect this to be the norm? It might be good for a support engineer to help debug the user's project to understand where the error code is coming from, and what exactly caused it.

**Comment by callmehiphop** [comment]
@bcoe FYI

**Comment by bcoe** [comment]
@callmehiphop aha! this does sound like the same thing; I'm seeing this crop up about `1,000,000` pub/sub requests in the `pub/sub` API.

**Comment by stephenplusplus** [comment]
@karthikv, we have pushed a fix recently that more eagerly refreshes an access token. Have you still been experiencing this issue using the latest dependencies?

**Comment by karthikv** [comment]
@stephenplusplus We'll try the new version (I assume you're referring to google-auth-library@5.3.0 or later) and report back if we see the same issue arise!

**Comment by stephenplusplus** [comment]
Yes, exactly. Thank you for linking that, and sounds good!

**Comment by eug48** [comment]
@stephenplusplus Thanks. I haven't seen this error since switching to the updated auth library, though it occurred quite infrequently for me. The update did take some fiddling though since both nodejs-common and nodejs-pubsub have "google-auth-library": "^5.0.0" - so yarn didn't update the library by default..

**Comment by karthikv** [comment]
@stephenplusplus We haven't experienced any issues in the past week of using the updated library. Thank you!
